{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Hello World\n"
     ]
    }
   ],
   "source": [
    "# Testing just to ensure this is working properly, as I've had issues with jupyter.\n",
    "print('Hello World')        # As per the norm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "    <div class=\"bk-root\">\n",
       "        <a href=\"https://bokeh.pydata.org\" target=\"_blank\" class=\"bk-logo bk-logo-small bk-logo-notebook\"></a>\n",
       "        <span id=\"7e7cfcb1-4a71-4e9d-8a51-3634bff137ca\">Loading BokehJS ...</span>\n",
       "    </div>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/javascript": [
       "\n",
       "(function(root) {\n",
       "  function now() {\n",
       "    return new Date();\n",
       "  }\n",
       "\n",
       "  var force = true;\n",
       "\n",
       "  if (typeof (root._bokeh_onload_callbacks) === \"undefined\" || force === true) {\n",
       "    root._bokeh_onload_callbacks = [];\n",
       "    root._bokeh_is_loading = undefined;\n",
       "  }\n",
       "\n",
       "  var JS_MIME_TYPE = 'application/javascript';\n",
       "  var HTML_MIME_TYPE = 'text/html';\n",
       "  var EXEC_MIME_TYPE = 'application/vnd.bokehjs_exec.v0+json';\n",
       "  var CLASS_NAME = 'output_bokeh rendered_html';\n",
       "\n",
       "  /**\n",
       "   * Render data to the DOM node\n",
       "   */\n",
       "  function render(props, node) {\n",
       "    var script = document.createElement(\"script\");\n",
       "    node.appendChild(script);\n",
       "  }\n",
       "\n",
       "  /**\n",
       "   * Handle when an output is cleared or removed\n",
       "   */\n",
       "  function handleClearOutput(event, handle) {\n",
       "    var cell = handle.cell;\n",
       "\n",
       "    var id = cell.output_area._bokeh_element_id;\n",
       "    var server_id = cell.output_area._bokeh_server_id;\n",
       "    // Clean up Bokeh references\n",
       "    if (id !== undefined) {\n",
       "      Bokeh.index[id].model.document.clear();\n",
       "      delete Bokeh.index[id];\n",
       "    }\n",
       "\n",
       "    if (server_id !== undefined) {\n",
       "      // Clean up Bokeh references\n",
       "      var cmd = \"from bokeh.io.state import curstate; print(curstate().uuid_to_server['\" + server_id + \"'].get_sessions()[0].document.roots[0]._id)\";\n",
       "      cell.notebook.kernel.execute(cmd, {\n",
       "        iopub: {\n",
       "          output: function(msg) {\n",
       "            var element_id = msg.content.text.trim();\n",
       "            Bokeh.index[element_id].model.document.clear();\n",
       "            delete Bokeh.index[element_id];\n",
       "          }\n",
       "        }\n",
       "      });\n",
       "      // Destroy server and session\n",
       "      var cmd = \"import bokeh.io.notebook as ion; ion.destroy_server('\" + server_id + \"')\";\n",
       "      cell.notebook.kernel.execute(cmd);\n",
       "    }\n",
       "  }\n",
       "\n",
       "  /**\n",
       "   * Handle when a new output is added\n",
       "   */\n",
       "  function handleAddOutput(event, handle) {\n",
       "    var output_area = handle.output_area;\n",
       "    var output = handle.output;\n",
       "\n",
       "    // limit handleAddOutput to display_data with EXEC_MIME_TYPE content only\n",
       "    if ((output.output_type != \"display_data\") || (!output.data.hasOwnProperty(EXEC_MIME_TYPE))) {\n",
       "      return\n",
       "    }\n",
       "\n",
       "    var toinsert = output_area.element.find(`.${CLASS_NAME.split(' ')[0]}`);\n",
       "\n",
       "    if (output.metadata[EXEC_MIME_TYPE][\"id\"] !== undefined) {\n",
       "      toinsert[0].firstChild.textContent = output.data[JS_MIME_TYPE];\n",
       "      // store reference to embed id on output_area\n",
       "      output_area._bokeh_element_id = output.metadata[EXEC_MIME_TYPE][\"id\"];\n",
       "    }\n",
       "    if (output.metadata[EXEC_MIME_TYPE][\"server_id\"] !== undefined) {\n",
       "      var bk_div = document.createElement(\"div\");\n",
       "      bk_div.innerHTML = output.data[HTML_MIME_TYPE];\n",
       "      var script_attrs = bk_div.children[0].attributes;\n",
       "      for (var i = 0; i < script_attrs.length; i++) {\n",
       "        toinsert[0].firstChild.setAttribute(script_attrs[i].name, script_attrs[i].value);\n",
       "      }\n",
       "      // store reference to server id on output_area\n",
       "      output_area._bokeh_server_id = output.metadata[EXEC_MIME_TYPE][\"server_id\"];\n",
       "    }\n",
       "  }\n",
       "\n",
       "  function register_renderer(events, OutputArea) {\n",
       "\n",
       "    function append_mime(data, metadata, element) {\n",
       "      // create a DOM node to render to\n",
       "      var toinsert = this.create_output_subarea(\n",
       "        metadata,\n",
       "        CLASS_NAME,\n",
       "        EXEC_MIME_TYPE\n",
       "      );\n",
       "      this.keyboard_manager.register_events(toinsert);\n",
       "      // Render to node\n",
       "      var props = {data: data, metadata: metadata[EXEC_MIME_TYPE]};\n",
       "      render(props, toinsert[0]);\n",
       "      element.append(toinsert);\n",
       "      return toinsert\n",
       "    }\n",
       "\n",
       "    /* Handle when an output is cleared or removed */\n",
       "    events.on('clear_output.CodeCell', handleClearOutput);\n",
       "    events.on('delete.Cell', handleClearOutput);\n",
       "\n",
       "    /* Handle when a new output is added */\n",
       "    events.on('output_added.OutputArea', handleAddOutput);\n",
       "\n",
       "    /**\n",
       "     * Register the mime type and append_mime function with output_area\n",
       "     */\n",
       "    OutputArea.prototype.register_mime_type(EXEC_MIME_TYPE, append_mime, {\n",
       "      /* Is output safe? */\n",
       "      safe: true,\n",
       "      /* Index of renderer in `output_area.display_order` */\n",
       "      index: 0\n",
       "    });\n",
       "  }\n",
       "\n",
       "  // register the mime type if in Jupyter Notebook environment and previously unregistered\n",
       "  if (root.Jupyter !== undefined) {\n",
       "    var events = require('base/js/events');\n",
       "    var OutputArea = require('notebook/js/outputarea').OutputArea;\n",
       "\n",
       "    if (OutputArea.prototype.mime_types().indexOf(EXEC_MIME_TYPE) == -1) {\n",
       "      register_renderer(events, OutputArea);\n",
       "    }\n",
       "  }\n",
       "\n",
       "  \n",
       "  if (typeof (root._bokeh_timeout) === \"undefined\" || force === true) {\n",
       "    root._bokeh_timeout = Date.now() + 5000;\n",
       "    root._bokeh_failed_load = false;\n",
       "  }\n",
       "\n",
       "  var NB_LOAD_WARNING = {'data': {'text/html':\n",
       "     \"<div style='background-color: #fdd'>\\n\"+\n",
       "     \"<p>\\n\"+\n",
       "     \"BokehJS does not appear to have successfully loaded. If loading BokehJS from CDN, this \\n\"+\n",
       "     \"may be due to a slow or bad network connection. Possible fixes:\\n\"+\n",
       "     \"</p>\\n\"+\n",
       "     \"<ul>\\n\"+\n",
       "     \"<li>re-rerun `output_notebook()` to attempt to load from CDN again, or</li>\\n\"+\n",
       "     \"<li>use INLINE resources instead, as so:</li>\\n\"+\n",
       "     \"</ul>\\n\"+\n",
       "     \"<code>\\n\"+\n",
       "     \"from bokeh.resources import INLINE\\n\"+\n",
       "     \"output_notebook(resources=INLINE)\\n\"+\n",
       "     \"</code>\\n\"+\n",
       "     \"</div>\"}};\n",
       "\n",
       "  function display_loaded() {\n",
       "    var el = document.getElementById(\"7e7cfcb1-4a71-4e9d-8a51-3634bff137ca\");\n",
       "    if (el != null) {\n",
       "      el.textContent = \"BokehJS is loading...\";\n",
       "    }\n",
       "    if (root.Bokeh !== undefined) {\n",
       "      if (el != null) {\n",
       "        el.textContent = \"BokehJS \" + root.Bokeh.version + \" successfully loaded.\";\n",
       "      }\n",
       "    } else if (Date.now() < root._bokeh_timeout) {\n",
       "      setTimeout(display_loaded, 100)\n",
       "    }\n",
       "  }\n",
       "\n",
       "\n",
       "  function run_callbacks() {\n",
       "    try {\n",
       "      root._bokeh_onload_callbacks.forEach(function(callback) { callback() });\n",
       "    }\n",
       "    finally {\n",
       "      delete root._bokeh_onload_callbacks\n",
       "    }\n",
       "    console.info(\"Bokeh: all callbacks have finished\");\n",
       "  }\n",
       "\n",
       "  function load_libs(js_urls, callback) {\n",
       "    root._bokeh_onload_callbacks.push(callback);\n",
       "    if (root._bokeh_is_loading > 0) {\n",
       "      console.log(\"Bokeh: BokehJS is being loaded, scheduling callback at\", now());\n",
       "      return null;\n",
       "    }\n",
       "    if (js_urls == null || js_urls.length === 0) {\n",
       "      run_callbacks();\n",
       "      return null;\n",
       "    }\n",
       "    console.log(\"Bokeh: BokehJS not loaded, scheduling load and callback at\", now());\n",
       "    root._bokeh_is_loading = js_urls.length;\n",
       "    for (var i = 0; i < js_urls.length; i++) {\n",
       "      var url = js_urls[i];\n",
       "      var s = document.createElement('script');\n",
       "      s.src = url;\n",
       "      s.async = false;\n",
       "      s.onreadystatechange = s.onload = function() {\n",
       "        root._bokeh_is_loading--;\n",
       "        if (root._bokeh_is_loading === 0) {\n",
       "          console.log(\"Bokeh: all BokehJS libraries loaded\");\n",
       "          run_callbacks()\n",
       "        }\n",
       "      };\n",
       "      s.onerror = function() {\n",
       "        console.warn(\"failed to load library \" + url);\n",
       "      };\n",
       "      console.log(\"Bokeh: injecting script tag for BokehJS library: \", url);\n",
       "      document.getElementsByTagName(\"head\")[0].appendChild(s);\n",
       "    }\n",
       "  };var element = document.getElementById(\"7e7cfcb1-4a71-4e9d-8a51-3634bff137ca\");\n",
       "  if (element == null) {\n",
       "    console.log(\"Bokeh: ERROR: autoload.js configured with elementid '7e7cfcb1-4a71-4e9d-8a51-3634bff137ca' but no matching script tag was found. \")\n",
       "    return false;\n",
       "  }\n",
       "\n",
       "  var js_urls = [\"https://cdn.pydata.org/bokeh/release/bokeh-0.12.10.min.js\", \"https://cdn.pydata.org/bokeh/release/bokeh-widgets-0.12.10.min.js\", \"https://cdn.pydata.org/bokeh/release/bokeh-tables-0.12.10.min.js\", \"https://cdn.pydata.org/bokeh/release/bokeh-gl-0.12.10.min.js\"];\n",
       "\n",
       "  var inline_js = [\n",
       "    function(Bokeh) {\n",
       "      Bokeh.set_log_level(\"info\");\n",
       "    },\n",
       "    \n",
       "    function(Bokeh) {\n",
       "      \n",
       "    },\n",
       "    function(Bokeh) {\n",
       "      console.log(\"Bokeh: injecting CSS: https://cdn.pydata.org/bokeh/release/bokeh-0.12.10.min.css\");\n",
       "      Bokeh.embed.inject_css(\"https://cdn.pydata.org/bokeh/release/bokeh-0.12.10.min.css\");\n",
       "      console.log(\"Bokeh: injecting CSS: https://cdn.pydata.org/bokeh/release/bokeh-widgets-0.12.10.min.css\");\n",
       "      Bokeh.embed.inject_css(\"https://cdn.pydata.org/bokeh/release/bokeh-widgets-0.12.10.min.css\");\n",
       "      console.log(\"Bokeh: injecting CSS: https://cdn.pydata.org/bokeh/release/bokeh-tables-0.12.10.min.css\");\n",
       "      Bokeh.embed.inject_css(\"https://cdn.pydata.org/bokeh/release/bokeh-tables-0.12.10.min.css\");\n",
       "    }\n",
       "  ];\n",
       "\n",
       "  function run_inline_js() {\n",
       "    \n",
       "    if ((root.Bokeh !== undefined) || (force === true)) {\n",
       "      for (var i = 0; i < inline_js.length; i++) {\n",
       "        inline_js[i].call(root, root.Bokeh);\n",
       "      }if (force === true) {\n",
       "        display_loaded();\n",
       "      }} else if (Date.now() < root._bokeh_timeout) {\n",
       "      setTimeout(run_inline_js, 100);\n",
       "    } else if (!root._bokeh_failed_load) {\n",
       "      console.log(\"Bokeh: BokehJS failed to load within specified timeout.\");\n",
       "      root._bokeh_failed_load = true;\n",
       "    } else if (force !== true) {\n",
       "      var cell = $(document.getElementById(\"7e7cfcb1-4a71-4e9d-8a51-3634bff137ca\")).parents('.cell').data().cell;\n",
       "      cell.output_area.append_execute_result(NB_LOAD_WARNING)\n",
       "    }\n",
       "\n",
       "  }\n",
       "\n",
       "  if (root._bokeh_is_loading === 0) {\n",
       "    console.log(\"Bokeh: BokehJS loaded, going straight to plotting\");\n",
       "    run_inline_js();\n",
       "  } else {\n",
       "    load_libs(js_urls, function() {\n",
       "      console.log(\"Bokeh: BokehJS plotting callback run at\", now());\n",
       "      run_inline_js();\n",
       "    });\n",
       "  }\n",
       "}(window));"
      ],
      "application/vnd.bokehjs_load.v0+json": "\n(function(root) {\n  function now() {\n    return new Date();\n  }\n\n  var force = true;\n\n  if (typeof (root._bokeh_onload_callbacks) === \"undefined\" || force === true) {\n    root._bokeh_onload_callbacks = [];\n    root._bokeh_is_loading = undefined;\n  }\n\n  \n\n  \n  if (typeof (root._bokeh_timeout) === \"undefined\" || force === true) {\n    root._bokeh_timeout = Date.now() + 5000;\n    root._bokeh_failed_load = false;\n  }\n\n  var NB_LOAD_WARNING = {'data': {'text/html':\n     \"<div style='background-color: #fdd'>\\n\"+\n     \"<p>\\n\"+\n     \"BokehJS does not appear to have successfully loaded. If loading BokehJS from CDN, this \\n\"+\n     \"may be due to a slow or bad network connection. Possible fixes:\\n\"+\n     \"</p>\\n\"+\n     \"<ul>\\n\"+\n     \"<li>re-rerun `output_notebook()` to attempt to load from CDN again, or</li>\\n\"+\n     \"<li>use INLINE resources instead, as so:</li>\\n\"+\n     \"</ul>\\n\"+\n     \"<code>\\n\"+\n     \"from bokeh.resources import INLINE\\n\"+\n     \"output_notebook(resources=INLINE)\\n\"+\n     \"</code>\\n\"+\n     \"</div>\"}};\n\n  function display_loaded() {\n    var el = document.getElementById(\"7e7cfcb1-4a71-4e9d-8a51-3634bff137ca\");\n    if (el != null) {\n      el.textContent = \"BokehJS is loading...\";\n    }\n    if (root.Bokeh !== undefined) {\n      if (el != null) {\n        el.textContent = \"BokehJS \" + root.Bokeh.version + \" successfully loaded.\";\n      }\n    } else if (Date.now() < root._bokeh_timeout) {\n      setTimeout(display_loaded, 100)\n    }\n  }\n\n\n  function run_callbacks() {\n    try {\n      root._bokeh_onload_callbacks.forEach(function(callback) { callback() });\n    }\n    finally {\n      delete root._bokeh_onload_callbacks\n    }\n    console.info(\"Bokeh: all callbacks have finished\");\n  }\n\n  function load_libs(js_urls, callback) {\n    root._bokeh_onload_callbacks.push(callback);\n    if (root._bokeh_is_loading > 0) {\n      console.log(\"Bokeh: BokehJS is being loaded, scheduling callback at\", now());\n      return null;\n    }\n    if (js_urls == null || js_urls.length === 0) {\n      run_callbacks();\n      return null;\n    }\n    console.log(\"Bokeh: BokehJS not loaded, scheduling load and callback at\", now());\n    root._bokeh_is_loading = js_urls.length;\n    for (var i = 0; i < js_urls.length; i++) {\n      var url = js_urls[i];\n      var s = document.createElement('script');\n      s.src = url;\n      s.async = false;\n      s.onreadystatechange = s.onload = function() {\n        root._bokeh_is_loading--;\n        if (root._bokeh_is_loading === 0) {\n          console.log(\"Bokeh: all BokehJS libraries loaded\");\n          run_callbacks()\n        }\n      };\n      s.onerror = function() {\n        console.warn(\"failed to load library \" + url);\n      };\n      console.log(\"Bokeh: injecting script tag for BokehJS library: \", url);\n      document.getElementsByTagName(\"head\")[0].appendChild(s);\n    }\n  };var element = document.getElementById(\"7e7cfcb1-4a71-4e9d-8a51-3634bff137ca\");\n  if (element == null) {\n    console.log(\"Bokeh: ERROR: autoload.js configured with elementid '7e7cfcb1-4a71-4e9d-8a51-3634bff137ca' but no matching script tag was found. \")\n    return false;\n  }\n\n  var js_urls = [\"https://cdn.pydata.org/bokeh/release/bokeh-0.12.10.min.js\", \"https://cdn.pydata.org/bokeh/release/bokeh-widgets-0.12.10.min.js\", \"https://cdn.pydata.org/bokeh/release/bokeh-tables-0.12.10.min.js\", \"https://cdn.pydata.org/bokeh/release/bokeh-gl-0.12.10.min.js\"];\n\n  var inline_js = [\n    function(Bokeh) {\n      Bokeh.set_log_level(\"info\");\n    },\n    \n    function(Bokeh) {\n      \n    },\n    function(Bokeh) {\n      console.log(\"Bokeh: injecting CSS: https://cdn.pydata.org/bokeh/release/bokeh-0.12.10.min.css\");\n      Bokeh.embed.inject_css(\"https://cdn.pydata.org/bokeh/release/bokeh-0.12.10.min.css\");\n      console.log(\"Bokeh: injecting CSS: https://cdn.pydata.org/bokeh/release/bokeh-widgets-0.12.10.min.css\");\n      Bokeh.embed.inject_css(\"https://cdn.pydata.org/bokeh/release/bokeh-widgets-0.12.10.min.css\");\n      console.log(\"Bokeh: injecting CSS: https://cdn.pydata.org/bokeh/release/bokeh-tables-0.12.10.min.css\");\n      Bokeh.embed.inject_css(\"https://cdn.pydata.org/bokeh/release/bokeh-tables-0.12.10.min.css\");\n    }\n  ];\n\n  function run_inline_js() {\n    \n    if ((root.Bokeh !== undefined) || (force === true)) {\n      for (var i = 0; i < inline_js.length; i++) {\n        inline_js[i].call(root, root.Bokeh);\n      }if (force === true) {\n        display_loaded();\n      }} else if (Date.now() < root._bokeh_timeout) {\n      setTimeout(run_inline_js, 100);\n    } else if (!root._bokeh_failed_load) {\n      console.log(\"Bokeh: BokehJS failed to load within specified timeout.\");\n      root._bokeh_failed_load = true;\n    } else if (force !== true) {\n      var cell = $(document.getElementById(\"7e7cfcb1-4a71-4e9d-8a51-3634bff137ca\")).parents('.cell').data().cell;\n      cell.output_area.append_execute_result(NB_LOAD_WARNING)\n    }\n\n  }\n\n  if (root._bokeh_is_loading === 0) {\n    console.log(\"Bokeh: BokehJS loaded, going straight to plotting\");\n    run_inline_js();\n  } else {\n    load_libs(js_urls, function() {\n      console.log(\"Bokeh: BokehJS plotting callback run at\", now());\n      run_inline_js();\n    });\n  }\n}(window));"
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Going to get around to make sure the standard things and the things Dr. X suggested are set up before we begin\n",
    "\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import urllib\n",
    "from Bio import SeqIO\n",
    "from Bio.Data import CodonTable\n",
    "from Bio.Seq import Seq\n",
    "from Bio.Alphabet import IUPAC\n",
    "from Bio.SeqUtils.ProtParam import ProteinAnalysis\n",
    "import numpy as np\n",
    "from bokeh.plotting import figure \n",
    "from bokeh.io import output_notebook, show\n",
    "output_notebook()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Function 1\n",
    "\n",
    "# First step, declaring the function.  Only one variable, fasta_fn.\n",
    "def get_sequences_from_file(fasta_fn):\n",
    "    sequence_data_dict = {}          # Creation of empty dictionary for our sequence data\n",
    "    for record in SeqIO.parse(fasta_fn, \"fasta\"):          #designates fasta filetype, returns iterated sequences with identifying info as SeqRecord objects\n",
    "        description = record.description.split()           #splits the description string into a number of substrings, stored as list\n",
    "        species_name = description[1] + \" \" + description[2]              #Takes second and third substrings from list, separates by space, stores as species name\n",
    "        sequence_data_dict[species_name] = record.seq          #adds the species name from above to the dictionary as a key mapped to the relevant sequence as a value\n",
    "    return(sequence_data_dict)          #returns completed dictionary of species names:sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#Function 2\n",
    "\n",
    "# Declare function.  Only one variable, that being the string of nucleotides.  So.\n",
    "def translate_function(string_nucleotides):\n",
    "    mito_table = CodonTable.unambiguous_dna_by_name[\"Vertebrate Mitochondrial\"]           #returns table of mitochondrial codons to amino acids\n",
    "    stop_codons = mito_table.stop_codons           #returns list of stop codons\n",
    "    sep_codons = [string_nucleotides[i:i+3] for i in range(0, len(string_nucleotides), 3)]           #Separates the codons into a list of individual codons\n",
    "    amino_seq = \"\"          #Initializes a blank string to amino_seq\n",
    "    for i in range(len(sep_codons)):\n",
    "        if len(sep_codons[i]) < 3:            #In case the sequence isn't divisible by three.  Worried about an error if that occurs, so added this.\n",
    "            print(amino_seq)\n",
    "            print(\"Sequence ends in stray nucleotides.\")\n",
    "            return\n",
    "        elif any(sep_codons[i] in s for s in stop_codons):              #Prints sequence and returns immediately if catches a stop codon.  If you want it to keep going after that point, things get iffy.\n",
    "            print(amino_seq)\n",
    "            return\n",
    "        else:\n",
    "            amino_seq+=mito_table.forward_table[sep_codons[i]]\n",
    "    print(amino_seq)\n",
    "    print(\"Sequence does not end in stop codon.\")\n",
    "    return"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "IV\n"
     ]
    }
   ],
   "source": [
    "translate_function('ATTGTTAGA')           #Testing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#Function 3\n",
    "\n",
    "#Yes, there is a better way to do it.  And it's pretty simple and straightforward, since you can use the translate function in biopython.\n",
    "\n",
    "def better_translate(string_nucleotides):\n",
    "    coding_dna = Seq(string_nucleotides)\n",
    "    translated_dna = coding_dna.translate(table=\"Vertebrate Mitochondrial\")            #Forgetting to add the correct table will result in you using standard genetic code \n",
    "    print(translated_dna)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "IV*\n"
     ]
    }
   ],
   "source": [
    "better_translate('ATTGTTAGA')                #For testing, again"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#Function 4\n",
    "\n",
    "def compute_molecular_weight(aa_seq):              #Must be string, yes.  Dunno why you wanted explicitly 3 amino acid sequence in the top, but I'm guessing it's a typo, since it doesn't seem to be what's wanted in the assignment proper.\n",
    "    fixed_seq = aa_seq.replace(\"*\",\"\")             #Removes the stop codons from above\n",
    "    fixed_seq = fixed_seq.replace(\" \",\"\")             #Also removed spaces, just in case \n",
    "    analysed_seq = ProteinAnalysis(fixed_seq)       #Just makes sure it's a protein sequence by building a sequence object.\n",
    "    return analysed_seq.molecular_weight()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "230.3039"
      ]
     },
     "execution_count": 79,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "compute_molecular_weight('IV*')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#Function 5\n",
    "\n",
    "#GC content analysis.\n",
    "\n",
    "def fraction_GC(string_nucleotides):\n",
    "    number_GC = string_nucleotides.count('G') + string_nucleotides.count('C')           #Pretty simple, just using count to determine number of G and adding to number of C\n",
    "    frac_GC = number_GC/len(string_nucleotides)          #Divide by total length of to get fraction GC\n",
    "    return frac_GC\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.4032258064516129"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fraction_GC('GACTTAATGGGCAATAGGCAAGCACTTGAAAAAGATGCCAACGACATGAAAACACAAGACAA')          #Testing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Start of main, just starting with what Dr. X has.\n",
    "\n",
    "cytb_seqs = get_sequences_from_file(\"penguins_cytb.fasta\") \n",
    "\n",
    "penguins_df = pd.read_csv(\"penguins_mass.csv\") # Includes only data for body mass \n",
    "species_list = list(penguins_df.species)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>species</th>\n",
       "      <th>mass</th>\n",
       "      <th>molecular_weight</th>\n",
       "      <th>GC_content</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Aptenodytes forsteri</td>\n",
       "      <td>28.00</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Aptenodytes patagonicus</td>\n",
       "      <td>13.40</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Eudyptes chrysocome</td>\n",
       "      <td>2.80</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Eudyptes chrysolophus</td>\n",
       "      <td>4.50</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Eudyptes sclateri</td>\n",
       "      <td>4.25</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Eudyptula minor</td>\n",
       "      <td>1.60</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Pygoscelis adeliae</td>\n",
       "      <td>4.60</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Pygoscelis antarctica</td>\n",
       "      <td>4.10</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Pygoscelis papua</td>\n",
       "      <td>6.10</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Spheniscus demersus</td>\n",
       "      <td>3.20</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Spheniscus humboldti</td>\n",
       "      <td>4.75</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Spheniscus magellanicus</td>\n",
       "      <td>3.40</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    species   mass  molecular_weight  GC_content\n",
       "0      Aptenodytes forsteri  28.00               NaN         NaN\n",
       "1   Aptenodytes patagonicus  13.40               NaN         NaN\n",
       "2       Eudyptes chrysocome   2.80               NaN         NaN\n",
       "3     Eudyptes chrysolophus   4.50               NaN         NaN\n",
       "4         Eudyptes sclateri   4.25               NaN         NaN\n",
       "5           Eudyptula minor   1.60               NaN         NaN\n",
       "6        Pygoscelis adeliae   4.60               NaN         NaN\n",
       "7     Pygoscelis antarctica   4.10               NaN         NaN\n",
       "8          Pygoscelis papua   6.10               NaN         NaN\n",
       "9       Spheniscus demersus   3.20               NaN         NaN\n",
       "10     Spheniscus humboldti   4.75               NaN         NaN\n",
       "11  Spheniscus magellanicus   3.40               NaN         NaN"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 6\n",
    "# Adding two new columns is pretty straightforward.\n",
    "\n",
    "penguins_df[\"molecular_weight\"] = np.nan            #Just imported numpy for this.  Pretty simple.  nan lists everything in that column as NaN.\n",
    "penguins_df[\"GC_content\"] = np.nan\n",
    "penguins_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MAPNLRKSHPLLKMINNSLIDLPTPSNISAWWNFGSLLGICLTTQILTGLLLAMHYTADTTLAFSSVAHTCRNVQYGWLIRNLHANGASFFFICIYLHIGRGFYYGSYLYKETWNTGIILLLTLMATAFVGYVLPWGQMSFWGATVITNLFSAIPYIGQTLVEWTWGGFSVDNPTLTRFFALHFLLPFMIAGLTLIHLTFLHESGSNNPLGIVANSDKIPFHPYYSTKDILGFALMLLPLTTLALFSPNLLGDPENFTPANPLVTPPHIKPEWYFLFAYAILRSIPNKLGGVLALAASVLILFLIPLLHKSKQRTMAFRPLSQLLFWALVANLIILTWVGSQPVEHPFIIIGQLASLTYFTTLLILFPIAGALENKMLNH*\n",
      "MAPNLRKSHPLLKMINNSLIDLPTPSNISAWWNFGSLLGICLTTQILTGLLLAMHYTADTTLAFSSVAHTCRNVQYGWLIRNLHANGASFFFICIYLHIGRGFYYGSYLYKETWNTGIILLLTLMATAFVGYVLPWGQMSFWGATVITNLFSAIPYIGQTLVEWAWGGFSVDNPTLTRFFALHFLLPFMIAGLTLIHLTFLHESGSNNPLGIVANSDKIPFHPYYSTKDTLGFALMLLPLTTLALFSPNLLGDPENFTPANPLVTPPHIKPEWYFLFAYAILRSIPNKLGGVLALAASVLILFLIPLLHKSKQRTMTFRPLSQLLFWTLVANLTILTWIGSQPVEHPFIIIGQLASLTYFTILLILFPLIGTLENKMLNH*\n",
      "MAPNLRKSHPLLKTINNSLIDLPTPSNISAWWNFGSLLGICLATQILTGLLLAAHYTADTTLAFSSVAHTCRNVQYGWLIRNLHANGASFFFICIYLHIGRGLYYGSYLYKETWNTGIILLLTLMATAFVGYVLPWGQMSFWGATVITNLFSAIPYIGQTLVEWAWGGFSVDNPTLTRFFTLHFLLPFMIAGLTLIHLTFLHESGSNNPLGIVANSDKIPFHPYYSTKDILGFILLLLPLTTLALFSPNLLGDPENFTPANPLVTPPHIKPEWYFLFAYAILRSIPNKLGGVLALAASVLILFLIPLLHKSKQRTMTFRPLSQLLFWTLVANLTILTWIGSQPVEHPFIIIGQLASLTYFTILLILFPLIGTLENKMLNH*\n",
      "MAPNLRKSHPLLKTINNSLIDLPTPSNISAWWNFGSLLGICLATQILTGLLLAAHYTADTTLAFSSVAHTCRNVQYGWLIRNLHANGASFFFICIYLHIGRGLYYGSYLYKETWNTGIILLLTLMATAFVGYVLPWGQMSFWGATVITNLFSAIPYIGQTLVEWAWGGFSVDNPTLTRFFTLHFLLPFMIAGLTLIHLTFLHESGSNNPLGIVANSDKIPFHPYYSTKDILGFILLLLPLTALALFSPNLLGDPENFTPANPLVTPPHIKPEWYFLFAYAILRSIPNKLGGVLALAASVLILFLIPLLHKSKQRTMTFRPLSQLLFWTLVANLTILTWIGSQPVEHPFIIIGQLASLTYFTILLILFPLIGTLENKMLNH*\n",
      "MAPNLRKSHPLLKTINNSLIDLPTPSNISAWWNFGSLLGICLATQILTGLLLAAHYTADTTLAFSSVAHTCRNVQYGWLIRNLHANGASFFFICIYLHIGRGLYYGSYLYKETWNTGIILLLTLMATAFVGYVLPWGQMSFWGATVITNLFSAIPYIGQTLVEWAWGGFSVDNPTLTRFFTLHFLLPFMIAGLTLIHLTFLHESGSNNPLGIVANSDKIPFHPYYSTKDILGFILLLLPLTTLALFSPNLLGDPENFTPANPLVTPPHIKPEWYFLFAYAILRSIPNKLGGVLALAASVLILFLIPLLHKSKQRTMTFRPLSQLLFWTLVANLTILTWIGSQPVEHPFIIIGQLASLTYFTILLILFPLIGTLENKMLNH*\n",
      "MAPNLRKSHPLLKMINNSLIDLPTPSNISTWWNFGSLLGICLITQILTGLLLAAHYTADTTLAFSSIAHTCRNVQYGWLIRNLHANGASFFFICIYLHIGRGLYYGSYLYKETWNTGIILLITLMATAFVGYVLPWGQMSFWGATVITNLFSAIPYIGQTLVEWAWGGFSVDNPTLTRFFALHFLLPFMITGLTLIHLTFLHESGSNNPLGIVANSDKIPFHPYYSTKDILGFILLLLPLTTLALFSPNLLGDPENFTPANPLVTPPHIKPEWYFLFAYAILRSIPNKLGGVLALAASVLILFLIPLLHKSKQRSMAFRPLSQLLFWTLVANLTILTWVGNQPVEHPFIIIGQLASLTYFTILLILFPLAGTLENKMLSH*\n",
      "MAPNLRKSHPLLKMINNSLIDLPTPSNISAWWNFGSLLGICLTTQILTGLLLAMHYTADTTLAFSSIAHTCRDVQYGWLIRNLHANGASFFFICIYLHIGRGLYYGSYLYKETWNTGIILLLTLMATAFVGYVLPWGQMSFWGATVITNLFSAIPYIGQTLVEWAWGGFSVDNPTLTRFFALHFLLPFVIMGLTLIHLTFLHESGSNNPLGIVANSDKIPFHPYYSTKDILGFAFMLLPLTTLALFSPNLLGDPENFTPANPLVTPPHIKPEWYFLFAYAILRSIPNKLGGVLALAASVLILFLIPLLHKSKQRTMAFRPLSQLLFWALVANLAILTWVGSQPVEHPFIIIGQLASLTYFTILLILFPSIGALENKMLNH*\n",
      "MAPNLRKSHPLLKMINNSLIDLPTPSNISAWWNFGSLLGICLTTQILTGLLLAMHYTADTTLAFSSIAHTCRDVQYGWLIRNMHANGASLLFICIYLHIGRGVYDGSYLHKETWNTGVILLLTLMATAFVGYVLPWGQMSFWGATVITNLFSAIPYIGQTIVEWAWGGFSVDNPTLTRFFALHFLLPFMITGLTLIHLTFLHESGSNNPLGIVANSDKIPFHPYYSTKDILGFTLMLLPLTTLALFSPNLLGDPENFTPANPLVTPPHIKPEWYFLFAYAILRSIPNKLGGVLALAASVLVLFLSPLLHKSKQRTMAFRPLSQLLFWTLVANLLILTWIGSQPVEHPFIIIGQLASTTYFIILLILFPITSALENKMLNF*\n",
      "MAPNLRKSHPLLKMINKSLIDLPTPPNISAWWNFGSLLGICLITQILTGLLLAMHYTADTTLAFSSIAHTCRNVQYGWLIRNLHANGASFFFICIYFHIGRGLYYGSYLYKETWNTGIILLLTLMATAFVGYVLPWGQMSFWGATVITNLFSAIPYVGQTLVEWAWGGFSVDNPTLTRFFALHFLLPFMITGLTLIHLTFLHESGSNNPLGIVANSDKIPFHPYYSTKDILGFALMLLPLTTLALFSPNLLGDPENFTPANPLVTPPHIKPEWYFLFAYAILRSIPNKLGGVLALAASVLILFLSPLLHKSKQRTMAFRPLSQLLFWTLVANLLILTWIGSQPVEHPFIIIGQLASTTYFIILLILFPITSALENKMLNF*\n",
      "MAPNLRKSHPLLKTINNSLIDLPTPSNISAWWNFGSLLGICLATQILTGLLLAAHYTADTTLAFSSVAHMCRNVQYGWLIRNLHANGASFFFICIYLHIGRGLYYGSYLYKETWNTGIILLLTLMATAFVGYVLPWGQMSFWGATVITNLFSAIPYIGQTLVEWAWGGFSVDNPTLTRFFALHFLLPFMITGLTLIHLTFLHESGSNNPLGIVANSDKIPFHPYYSTKDILGFMLLLLPLTTLALFSPNLLGDPENFTPANPLVTPPHIKPEWYFLFAYAILRSIPNKLGGVLALAASVLILFLMPLLHKSKQRTMAFRPLSQLLFWTLVANLIILTWIGSQPVEHPFIAIGQLASFTYFTTLLILFPLAGALENKMLNH*\n",
      "MAPNLRKSHPLLKTINNSLIDLPTPSNISAWWNFGSLLSICLATQILTGLLLAAHYTADTTLAFSSVAHMCRNVQYGWLIRNLHANGASFFFICIYLHIGRGLYYGSYLHKETWNTGIILLLTLMATAFVGYVLPWGQMSFWGATVITNLFSAIPYIGQTLVEWAWGGFSVDNPTLTRFFALHFLLPFMITGLALIHLTFLHESGSNNPLGIVANSDKIPFHPYYSTKDILGFMLLLLPLTTLALFSPNLLGDPENFTPANPLVTPPHIKPEWYFLFAYAILRSIPNKLGGVLALAASVLILFLMPLLHKSKQRTMAFRPLSQLLFWTLVANLIILTWVGSQPVEHPFIIIGQLASLTYFTTLLILFPLAGALENKMLNH*\n",
      "MAPNLRKSHPLLKTINNSLIDLPTPSNISAWWNFGSLLGICLATQILTGLLLAAHYTADTTLAFSSVAHMCRNVQYGWLIRNLHANGASFFFICIYLHIGRGLYYGSYLYKETWNTGIILLLTLMATAFVGYVLPWGQMSFWGATVITNLFSAIPYIGQTLVEWAWGGFSVDNPTLTRFFALHFLLPFMITGLTLIHLTFLHESGSNNPLGIVANSDKIPFHPYYSTKDILGFMLLLLPLTTLALFSPNLLGDPENFTPANPLVTPPHIKPEWYFLFAYAILRSIPNKLGGVLALAASVLILFLMPLLHKSKQRTMAFRPLSQLLFWTLVANLIILTWVGSQPVEHPFIIIGQLASFTYFTTLLILFPLAGALENKMLNH*\n"
     ]
    }
   ],
   "source": [
    "#7\n",
    "\n",
    "\n",
    "for key, value in cytb_seqs.items():            #Can't figure out how to get this to work.  Can't even figure out if it's pulling the right things.\n",
    "    a_seq = better_translate(str(value))\n",
    "    molec = compute_molecular_weight(str(a_seq))\n",
    "    GC_dec = fraction_GC(str(value))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0        Aptenodytes forsteri\n",
       "1     Aptenodytes patagonicus\n",
       "2         Eudyptes chrysocome\n",
       "3       Eudyptes chrysolophus\n",
       "4           Eudyptes sclateri\n",
       "5             Eudyptula minor\n",
       "6          Pygoscelis adeliae\n",
       "7       Pygoscelis antarctica\n",
       "8            Pygoscelis papua\n",
       "9         Spheniscus demersus\n",
       "10       Spheniscus humboldti\n",
       "11    Spheniscus magellanicus\n",
       "Name: species, dtype: object"
      ]
     },
     "execution_count": 118,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "penguins_df['species']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAaQAAAEICAYAAAAQkoCgAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAIABJREFUeJzt3Xm8XHV5x/HPNwuBEFYTNo1EARdaFTCCuBVUFLRuFRRE\nBKuirdTdFq2tuBXcq6IoKoKiqFVQilRFZBE3CBggEASEsAYIJJCdJDdP/3iecYbLvTeTZe49l3zf\nr9e87syc81vPOb/n/M6cmauIwMzMbKSNGekKmJmZgQOSmZk1hAOSmZk1ggOSmZk1ggOSmZk1ggOS\nmZk1ggOSrTOlb0laIOnSka7PYCRdKOnNI12PgUg6XNIvh7nMV0m6TdJiSXsOZ9nrStJjq75jR7ou\n1jsOSKOQpDmSVkia3O/9mZJC0rRhqspzgAOAx0TE3sNU5iNKRHw3Il40zMV+BjgmIiZFxJ+Guex1\nEhG3Vn37Rrou1jsOSKPXzcBhrReSngJsNsx12BmYExFL1jahpHE9qI91Z2fgmpGuhFl/Dkij13eA\nN3S8PhL4ducKkl4q6U+SFtYlmuM6lm0q6XRJ90m6X9JlkravZUdJuknSIkk3Szq8f+GS3gR8A9i3\nLqV8pN5/i6QbJc2XdLaknTrShKS3S7oBuGGgRkl6pqTfVZ2ulLRfx7I3Sppd9bpJ0lv7pX1FzRIX\nSvqLpAM7Fu8s6beV9pf9Z5cdeUyWdE6VP1/SbySNqWVzJH1A0rV1mfJbkjbtSPv3Vf791Yandiyb\nKulMSfOqz0/s6OtLOtZ7kqTzquw/S3pNx7KXVNmLJN0h6X2DtGGMpA9JukXSPZK+LWkrSRMkLQbG\nAldK+ssg6UPSO6qP75X06VYf1PJ/rO2wQNIvJO3cL+3bJN1Qy78sSbVsrKTPVp43Szqm1h/X0b8v\n7MjrOEmn1/Np/da9UNLHutmmNopEhB+j7AHMAV4I/Bl4MjnA3Eae+QYwrdbbD3gKeeLxVOBu4JW1\n7K3A/wITK/3TgS2BzYGFwBNrvR2BvxmkHkcBl3S8fj5wL7AXMAH4EnBxx/IAzgO2BTYbIL9HA/cB\nL6k6H1Cvp9TylwK7AAL+DlgK7FXL9gYeqDRjKq8n1bILgb8ATyBnkRcCJwzSpuOBrwLj6/FcQB39\nPguYWm34LfDxWrYXcA+wT/XnkbX+hHp9JfD56t9Ngef078NadhvwRmBc5Xlvq/+BucBz6/k2rbYP\n0IZ/BG4EHg9MAs4EvtNvO+w6xP4VwAXVxscC1wNvrmWvrLyfXHX8EPC7fmnPAbautPOAA2vZ24Br\ngcdU/X9V64/r3K878joOOL2eT+u3btfb1I/R8/AMaXRrzZIOAK4D7uhcGBEXRsTVEbE6Iq4CziAH\ncoCVwKPIgakvIi6PiIW1bDXwt5I2i4i5EdHt5Z3DgVMi4oqIeBD4ADmDmtaxzvERMT8ilg2Q/vXA\nuRFxbtX5PGAGGaCIiJ9FxF8iXQT8kgwYAG+qss+rtHdExHUdeX8rIq6vcn8I7DFIG1aSQXjniFgZ\nEb+JiM4ffDwxIm6LiPnAJ2hfNn0L8LWI+GP152nAg8AzyWC5E/D+iFgSEcsj4hIe7u/JS6DfiohV\nEXEF8GPg4I667S5py4hYUMsHcjjwuYi4KSIWk9vhUK3dZdJP1na6Ffjvjna+ldyGsyNiFfBfwB6d\nsyQyMNxfaS+g3devAb4QEbdHxALghLWoz0C63aY2SjggjW7fAV5HnmV/u/9CSftIuqAuEz1AnqFO\n7kj7C+D7ku6U9ClJ4yM/D3ptrTtX0s8kPanL+uwE3NJ6UYPhfeRspeW2IdLvDBxSl7zul3Q/eePE\njtWegyT9oS5n3U8GqlZ7ppJnzIO5q+P5UnLmMJBPkzOAX9Ylq2P7Le+s/y1km1t1f2+/uk+t5VOB\nW2oAH8rOwD798jgc2KGWv5ps8y2SLpK07yD5PGQ71PNxwPZrKL/TUO38Qkf95pMz1s5tPFhf79Qv\n36H2hW50u01tlHBAGsUi4hby5oaXkJdl+vsecDYwNSK2Ii9FqdKujIiPRMTuwLPIs/M31LJfRMQB\nZCC4Dvh6l1W6kxywAJC0OTkL65y5DfXz8reRl5a27nhsHhEnSJpAzhY+A2wfEVsD57baU2l36bKe\ng4qIRRHx3oh4PPAy4D2SXtCxytSO548l29wq/xP96j4xIs6oZY/tYoZyG3BRvzwmRcQ/Vd0ui4hX\nANsBPyFnBQN5yHaoeq4iL9l2a6h2vrVfHTeLiN91kedc8nLdQGUALCEvIbfsgG1UHJBGvzcBz4+B\n73TbApgfEcsl7U3OpgCQtL+kpyi/17GQvBzUJ2l7SS+vYPIgsBjo9lbb7wFvlLRHBZD/Av4YEXO6\nTH868DJJL64PwDeVtJ+kxwCbkJ/HzANWSToI6Lxd+ptV9gvqQ/1Hr8XM7q+UNybsWh/ELyTb3tn+\nt0t6jKRtgQ8CP6j3vw68rWalkrS58qaSLYBLycH4hHp/U0nPHqD4c4AnSDpC0vh6PEPSkyVtovzO\n0lYRsbKjbgM5A3i3pMdJmkRuhx90MUPr9H5J20iaCryzo51fBT4g6W+qv7aSdEiXef4QeGdtm62B\nf+u3fCZ5aXG8pOm0L1XaRsIBaZSrz1RmDLL4n4GPSloE/CcPPaPeAfgRObDNBi4iA8IY4L3kGfF8\n8jOnf+6yLucD/0HOZOaSM5ZD16IttwGvIAf6eeTZ+PuBMRGxCHhHtWEBGVzP7kh7KXkzwOfJmxsu\n4qGzhG7tRn7Yvhj4PfCViLiwY/n3yM+ubqrHx6v8GeTnSCdW/W4kL6US+d2ZlwG7ArcCt5OXRfu3\nfxEZZA8l+/8u4JNkIAY4ApgjaSF5SfX1g7ThFPKS7MXkDHo58C9r0QcAPwUuJ4PEz8iAT0ScVXX6\nftVjFnBQl3l+ney7q4A/kTPcVbQD63+Q+8wC4CNkX9tGpHX3kJmtgaQ55N1mvxrpuvSSpAB2i4gb\ne1zOQcBXI2JdThzsEcgzJDMbFpI2U36XapykRwMfBs4a6XpZczggmdlwEXkpbgF5yW42eSnZDPAl\nOzMzawjPkMzMrBFGxQ9cTp48OaZNmzbS1TAzG1Uuv/zyeyNiykjXo1ujIiBNmzaNGTMGu7PZzMwG\nIumWNa/VHL5kZ2ZmjeCAZGZmjeCAZGZmjeCAZGZmjeCAZGZmjeCAZGZmjeCAZGZmjeCAZGZmjeCA\nZGZmjTAqfqlhvVxw/IbPc/8PbPg8zcw2cp4hmZlZIzggmZlZIzggmZlZIzggmZlZIzggmZlZIzgg\nmZlZIzggmZlZIzggmZlZIzggmZlZIzggmZlZIzggmZlZIzggmZlZIzggmZlZIzggmZlZIzggmZlZ\nIzggmZlZIzggmZlZIzggmZlZI/QsIEmaKukCSbMlXSPpnfX+cZLukDSzHi/pVR3MzGz0GNfDvFcB\n742IKyRtAVwu6bxa9vmI+EwPyzYzs1GmZwEpIuYCc+v5IkmzgUf3qjwzMxvdhuUzJEnTgD2BP9Zb\nx0i6StIpkrYZJM3RkmZImjFv3rzhqKaZmY2gngckSZOAHwPvioiFwEnALsAe5AzqswOli4iTI2J6\nREyfMmVKr6tpZmYjrKcBSdJ4Mhh9NyLOBIiIuyOiLyJWA18H9u5lHczMbHTo5V12Ar4JzI6Iz3W8\nv2PHaq8CZvWqDmZmNnr08i67ZwNHAFdLmlnvfRA4TNIeQABzgLf2sA5mZjZK9PIuu0sADbDo3F6V\naWZmo5d/qcHMzBrBAcnMzBrBAcnMzBrBAcnMzBrBAcnMzBrBAcnMzBrBAcnMzBrBAcnMzBrBAcnM\nzBrBAcnMzBrBAcnMzBrBAcnMzBrBAcnMzBrBAcnMzBrBAcnMzBrBAcnMzBrBAcnMzBrBAcnMzBrB\nAcnMzBrBAcnMzBrBAcnMzBrBAcnMzBrBAcnMzBrBAcnMzBrBAcnMzBrBAcnMzBrBAcnMzBrBAcnM\nzBqhZwFJ0lRJF0iaLekaSe+s97eVdJ6kG+rvNr2qg5mZjR69nCGtAt4bEU8Gngm8XdLuwLHA+RGx\nG3B+vTYzs41czwJSRMyNiCvq+SJgNvBo4BXAabXaacAre1UHMzMbPYblMyRJ04A9gT8C20fEXMig\nBWw3SJqjJc2QNGPevHnDUU0zMxtBPQ9IkiYBPwbeFRELu00XESdHxPSImD5lypTeVdDMzBqhpwFJ\n0ngyGH03Is6st++WtGMt3xG4p5d1MDOz0aGXd9kJ+CYwOyI+17HobODIen4k8NNe1cHMzEaPcT3M\n+9nAEcDVkmbWex8ETgB+KOlNwK3AIT2sg5mZjRI9C0gRcQmgQRa/oFflmpnZ6ORfajAzs0ZwQDIz\ns0ZwQDIzs0ZwQDIzs0ZwQDIzs0ZwQDIzs0ZwQDIzs0ZwQDIzs0ZwQDIzs0ZwQDIzs0ZwQDIzs0Zw\nQDIzs0ZwQDIzs0ZwQDIzs0ZwQDIzs0ZwQDIzs0ZwQDIzs0ZwQDIzs0ZwQDIzs0ZwQDIzs0ZwQDIz\ns0ZwQDIzs0ZwQDIzs0ZwQDIzs0boKiBJOkTSFvX8Q5LOlLRXb6tmZmYbk25nSP8REYskPQd4MXAa\ncFLvqmVmZhubbgNSX/19KXBSRPwU2KQ3VTIzs41RtwHpDklfA14DnCtpwlqkNTMzW6Nug8prgF8A\nB0bE/cC2wPuHSiDpFEn3SJrV8d5xku6QNLMeL1nnmpuZ2SNKtwFpR+BnEXGDpP2AQ4BL15DmVODA\nAd7/fETsUY9zu66pmZk9onUbkH4M9EnaFfgm8Djge0MliIiLgfnrVz0zM9tYdBuQVkfEKuAfgP+O\niHeTs6Z1cYykq+qS3jaDrSTpaEkzJM2YN2/eOhZlZmajRbcBaaWkw4A3AOfUe+PXobyTgF2APYC5\nwGcHWzEiTo6I6RExfcqUKetQlJmZjSbdBqQ3AvsCn4iImyU9Djh9bQuLiLsjoi8iVgNfB/Ze2zzM\nzOyRaVw3K0XEtcA7Ol7fDJywtoVJ2jEi5tbLVwGzhlrfzMw2Hl0FJEm7AccDuwObtt6PiMcPkeYM\nYD9gsqTbgQ8D+0naAwhgDvDWda24mZk9snQVkIBvkQHl88D+5CU8DZUgIg4b4O1vrlXtzMxso9Ht\nZ0ibRcT5gCLilog4Dnh+76plZmYbm25nSMsljQFukHQMcAewXe+qZWZmG5tuZ0jvAiaSNzY8HTgC\nOLJXlTIzs41Pt3fZXVZPF5OfH5mZmW1QQwYkSWcPtTwiXr5hq2NmZhurNc2Q9gVuA84A/sga7qwz\nMzNbV2sKSDsABwCHAa8DfgacERHX9Lpio84Fx2/4PPf/wIbP08ysoYa8qaF+5ufnEXEk8EzgRuBC\nSf8yLLUzM7ONxhpvaqj/DvtScpY0DfgicGZvq2VmZhubNd3UcBrwt8D/AR+JCP/2nJmZ9cSaZkhH\nAEuAJwDvkP56T4OAiIgte1g3MzPbiAwZkCKi2y/OmpmZrRcHHDMzawQHJDMzawQHJDMzawQHJDMz\nawQHJDMzawQHJDMzawQHJDMzawQHJDMzawQHJDMzawQHJDMzawQHJDMzawQHJDMzawQHJDMzawQH\nJDMzawQHJDMzawQHJDMza4SeBSRJp0i6R9Ksjve2lXSepBvq7za9Kt/MzEaXXs6QTgUO7PfescD5\nEbEbcH69NjMz611AioiLgfn93n4FcFo9Pw14Za/KNzOz0WW4P0PaPiLmAtTf7QZbUdLRkmZImjFv\n3rxhq6CZmY2Mxt7UEBEnR8T0iJg+ZcqUka6OmZn12HAHpLsl7QhQf+8Z5vLNzKyhhjsgnQ0cWc+P\nBH46zOWbmVlD9fK27zOA3wNPlHS7pDcBJwAHSLoBOKBem5mZMa5XGUfEYYMsekGvyjQzs9GrsTc1\nmJnZxsUByczMGsEByczMGsEByczMGsEByczMGsEByczMGsEByczMGsEByczMGsEByczMGsEByczM\nGsEByczMGsEByczMGsEByczMGsEByczMGsEByczMGsEByczMGsEByczMGsEByczMGsEByczMGsEB\nyczMGsEByczMGsEByczMGsEByczMGmHcSFfAzBrqguM3bH77f2DD5mePOJ4hmZlZIzggmZlZIzgg\nmZlZIzggmZlZI4zITQ2S5gCLgD5gVURMH4l6mJlZc4zkXXb7R8S9I1i+mZk1iC/ZmZlZI4xUQArg\nl5Iul3T0QCtIOlrSDEkz5s2bN8zVMzOz4TZSAenZEbEXcBDwdknP679CRJwcEdMjYvqUKVOGv4Zm\nZjasRuQzpIi4s/7eI+ksYG/g4pGoi5ltBDb0r06Af3miB4Y9IEnaHBgTEYvq+YuAjw53Pcw2OA96\nZutlJGZI2wNnSWqV/72I+PkI1MOG4t8xM7NhNuwBKSJuAp423OVaA3lGYWYd/GvfZqONZ6/2COXv\nIZmZWSM4IJmZWSM4IJmZWSM4IJmZWSM4IJmZWSP4Ljszsw3FX2VYLw5IZjZyPIBbB1+yMzOzRnBA\nMjOzRvAlu9HG39I3s0coz5DMzKwRHJDMzKwRHJDMzKwRHJDMzKwRHJDMzKwRHJDMzKwRHJDMzKwR\nHJDMzKwR/MVYe+Tz76WZjQqeIZmZWSM4IJmZWSM4IJmZWSM4IJmZWSM4IJmZWSM4IJmZWSM4IJmZ\nWSM4IJmZWSOMSECSdKCkP0u6UdKxI1EHMzNrlmEPSJLGAl8GDgJ2Bw6TtPtw18PMzJplJGZIewM3\nRsRNEbEC+D7wihGoh5mZNYgiYngLlA4GDoyIN9frI4B9IuKYfusdDRxdL58I/LnHVZsM3NvjMlxO\ns8t5JLXF5TS3jOEsZ+eImDIM5WwQI/HjqhrgvYdFxYg4GTi599VJkmZExHSXs/GW80hqi8tpbhnD\nWc5oMxKX7G4Hpna8fgxw5wjUw8zMGmQkAtJlwG6SHidpE+BQ4OwRqIeZmTXIsF+yi4hVko4BfgGM\nBU6JiGuGux4DGK7Lgy6nueU8ktricppbxnCWM6oM+00NZmZmA/EvNZiZWSM4IJmZWSNskIAk6VWS\nQtKTulz/g+uTbi3qtXgNy/eQ1CdpZsfj2C7TvaSez5E0WdIhkmZLuqBjvYfl3U17OvpnjqTJ/Za9\ncrBftpD0jdYySZ+SNGuodnRTx/751utprbwHquNalPXzLtMdJenEQfqjlecsSf8jaeIgaVt9eqGk\nAW+3rbzmS1og6cz+eXVRz49KeuEA78+RdJ6krSX9bi3ye1jbOvtA0u86t0VHun+XdI2kqyr9Kzu2\n13RJX+y3/qn1/cDO9/66jw9St1YZKyRdLWmfqs9gfXCJpLd0vP6GpN0HqOs+a+iTU6svHlZGLR90\n+w6w7oB1HWC9oySd2E2ea2tNY80A6z9sW9X7+0k6p+P5s1r5S3qbpPd0Mx4MUe5DxoCeiYj1fgA/\nBH4DHNfl+ovXJ91a1GvI9YGjgBXrmO7Eej6H/JLbz4H9u82HvKFkwOUd/TMHmNxv2anAwQOkGdvv\n9RJg1obu00ozrZX3QHXsbONQZXVbdqu/B+mPxR3Pvwu8Z5C0rT69EJg+WF9U/94D/Kh/Xmuo49gh\nlg3aR91um1bb+ufVuS3q9b7A74EJ9Xoy+esog+4LA+1Tnfv4AOv/tYyqzxOAnfpv735p7gL+dbB8\nOuq60xr6ZMD9v2P5oNt3XR+D9QX5ncox65n32o5pgx3/+wHn1PPjgPd15t9/P2nqY0NsrEnAHbVT\nXtevgy4GzgKuBb5KzshOAPqAq4Clle5O4FJgJnkLeCvdauAK4Moqo6/W+QlwPnAr8ABwDfA1YJfa\nwfvIQNcH/AH4H/LniXaudAs68oyq3/XA5ZV/H/AgcEmlew2wEJhfZS4HVpLftJ4HfKbSrADuBg6u\n/FcDy+rxQC37S7X7tkoTHevNqXKj8uoDZpHBZUUtWwncT36faxlwH7Cq8lgN3AB8p9KurrreU68v\nq/YtqnovrnVWADcBN1bfrqxlHyB/IWNVPa4EPlbPo/JcCdxMDgR/rHSrql6/6GhHq62LgE068mjV\n+4Aqq1Xvu4H9q09a760CPl3l3VHbcgnw4+rTZbXeg7WNZ1Y/t8pq9d8ttT1X1/a4q8qJymd1rdfX\n0faVVafZ5AF/Vke9VlS7Hqzt8kClWQCcUnnfCnyhY5tcW3VfVHW5sdo/izw2ZtS6S6vsW6vNy8gA\ne2WV//uq18Xk/tiq80Lga3Usvq7yaPXNg+TxdmZHG/qqXnfVstb2WkXuP3dVuqXVviXA1zuW/6ye\nLwau67fN7qrXCyr9klq2qB730j4G7qrt01mv1vFwCzC36nEXuc9eUm1bVvVYWX9nk8drK5/F1b5f\nVj6tfecvwKcq/1Z/f6b67SNV5xWV9g+096evAH+qusys7Ta/6nFV/f0RcAG5/efWNlrSsQ3uqL93\n1/sXAVOAt9A+7h8gx8RtyPFtHrmvfLTKuK764BJyXJhdbb2j6rWU3F8/Qfskcho5Pl5Rj2d1jNkX\nVr2vI0+CWje+XUgFeuBA2uPy+f2DYL2eVeVsTu4bV9Z7rx0qnmyIS3avBH4eEdcD8yXt1bFsb+C9\nwFPIzvyHiDiW7MhPAWeQt35vChwTEXuQO8UzK53q8bHqnFW1zljgvGrk+8hBuI8MPCeRgW9WlXMx\nuVHfSB7I/0NuuC+RO0IAjybP9h5Ty5bVsscBbwZeTe5QDwD/Sw60XwOOJM/qxpKD5E3kzxydBDy+\n+mBJtW8u8NMqaymwJ7ljQ2786cBW1aYl5A4whgx2Au6OiAlVjoCnVb5bkrOz95I797XAMdWuqL78\nSqU5sfpyPHnQvKreXwJsVu1dDHyRnMG9r9abW/35IuDfaA+G36z1HiTPkB9HDqQPVBvPrTb0Aa3L\nspuSA+9YYGVEjCEP9o+TgeonlfYPlb41uPxfpXlWteNucp+YAPw9uT8sr7KOAraoOp1NO9AcVfX5\nCHkSpGrPF8gva/eRs3aqn4+r9ZaR+827gO3IE5T7q91XkINcK9BsW+v/oNaZUttoDPDSWjYO+DXw\nX+SAdza5PzwI/Cvw2krTFxETqw+2I/ef+4A3Ae+ovJ5c7bu9ttPx5MnV5sChkl5P7kO3djyW1XZ6\nMTmQXVv9PaP6+Mm1Hf4X+F6Vfx05iB9InrhNJPeHleR+e3a1fyx5HH2mttH5wG5k0Lm88mxdMvw+\nOYhvW31/EHlyeQ65T1J98tnaVmOAX1W+44B3krOsueSJ4++r/Htq/TuBp9a2uavqd3/V/X5y3JgN\nvJsMUttXuR+X9CLg9eS+8taq36fJY34sOebtWeVsERF/W/X9aUQ8lTwh2AJ4PvChyvuX1V9zqx8n\nkfv8e8jbwJcAHyZPFFYDLyHHqfH1/hdqWx1bbZwAvIwMTjvWtvybavdPOsbT/u4BDoiIvch9rfMS\n7p7kfr47OYY9uzOhpCnkicirI+JpwCED5N/pQODOiHha9dGQl+k3REA6jNyxqL+HdSy7NPJHVPvI\n4POcAdK9gNzA50iaCewFzI2Im8gzky9Vuss76rsvOcA8HXg7OSC9gBz0zqh0/1brXl5/dyUHszHk\nGfVp5IHSOgPelTz72Y4cnHcgD/QnkWdR08iDeX9ywxMRPyMHlP3JHWIqOUvYgjyAO8+c3wf8jjzo\nN6k69dXfieTOP4k8gDarcleTM8rNgJ0k3UYGtIURcV8tv5YcGN5U/bh7RDxQz2+KiJlkkFHldVI9\nv7XqQuVxHzlI7gJ8q+o+F3g57bPUvav/PgncHBFHVx3uJQ+466sNK4Cr629rptHqz0vJgTWAMTVg\nriAHiMeQB8Tm1f5NK90TyQM7yJObL1Z9r652Qh68C4EZEXFG1euPZEC+rdLuQw7Gk2nPKj9FDkiq\nvF5e+U2tx/bVpkPIQDieduBeSgauM+rvprV8EnmCsZCclU6gHbA2IQeEM8mTpB1qvTHA1hFxLrkv\nbwWMk9RHBtaxtK8SqNKMrfyXVvu3BY6o/lxWy79cz8eT++iW5P55Z9Vr79rmj622blvteDbwQvKL\n608ij8EXkwHh2dWeI1vbkdwHqTJ/R57EiTzrPqDWu7C243G17Bzy+FxBBpjzgD0q7eaV3ybkjGEc\necLXR844ZlWbxlQ73k6epPWRAXWratdltc7OldfllVfru4+3kMFqCnnSocr3ReQx/0YyKK0Ctq66\nBhlIqefbSPoSORY8s/ZpgN9EThdmV71uJ09gtiZPjlozxR+Q+/Jd1c/7kFeOTgQOr3yfR457cyrv\nPwERETdUXScBz638J1d7BjMe+Lqkq6vNnZ8NXRoRt0fEanKGNa1f2mcCF0fEzWQF5jO0q4EXSvqk\npOfW2DSo9QpIkh5F7mDfkDQHeD/wWkmt36vr/yWnztfPB75BbmyRB+uewBtod/rKjnR9PJTIoPIM\nYEFEPJE88KIjHZVuHHkZawK5I3xrgPpE5fn6yudWMjh+g9zYK4Hfkjv5LgN0xyzyLP/n5AD9Y3In\n/vfK+6vkgfIguROOoX256kHy4F5OBr9lVe/FwH9XntfRvuzy7SpzBbnjLiaD8320B2gq31Zf9ZGz\nuu+Rg+eXI2JB1fFicpB9KvlZyKwq5+Sqy2ZV94HypvIQObvYgRwMfkgGtN/Xsi/X38W0Zywrq95b\nVP3mk2d/q2t7Bnmp4RhyMFtFO5h/g9xXIGd6vyADz561Xy4nt9NOtAcfdfydQHu29EPal2zOJmd4\nfyIvdS0mZ3T3V5pDaQf4VttbfXFvPa6o+v+pX5+9qNq5GXnW/Lxq46sr3xaRQRtyQLiI3GfHd/Th\nUeR+fVb15Y9o7x+fI8/GjyJnxI+qfriHPLtdSvuy6fyq89JqX6v806vd/0IG7GuBd0fEJmTA7IuI\nCyvdHKA1Q6DKeV6VPY6cAWxTfdypFcyurX45rN77Vb1+gDz5++da/xm092VoX45bQJ5QLq80y8nZ\nk6qu15IntktoB/TWjwIEuc1fU/lsS54sqNq2N/CfwHci4pu0L/92/qjAV8n984F6/+nV563ZyXja\nY8215KxxFQ8NGqsrbZBXNBZExFPI/XMCQxO5/X9Ts6Ivkid+g3k3Oct8GnnitEnHss7jujV29i9r\noC+vruJFpafbAAAGEElEQVSh8WRTgLpy9nQyMB0v6T+Hasj6zpAOBr4dETtHxLSImEoOdq2Z0N7K\nnwgaQ04NL6n3BZweETuTUf1u8uzhOeTgtI+kx9W6nelWSxpPnoFtWuW/DbhE0rbk2UHrwD68X11P\nrfZuG/nLEIeTZy4iA82h5KD2uVp/DrmTngr8IzlIfpDcoZ4BbCHpIPJAu4A8u7+A9lT8CbQvM0wg\nd8Kp5KzsVnIj9dG+pHUaOVAtIweWR5Fns9cAf1d/x9C+3NCyKXkWeAx5ZnSfpC3JHWTLWmdipd2c\nDD7bkzOuJ5ID5hfIs9UptIP1InI28vOq/9iqw2ryxANJx1W+95E78q5Vh83Is6s55NnXWPIMnCrj\nt7RPQj5I+zOCCR15v4wcEN5DHhQryQN7RW2rs8jtD3np5d5abxPyMnJfbYOJ1d+ryOC4SbWtNRhM\nJmfOnQf9JuQ19qXk4Lo1GWQnVv9dWn05sdqxPRk4Wpeepkl6eq3zvGqnanv0kWfCe1Xe15JB4cnA\nAkkvJveVVtvurbq1ZiCrK92rql+nVb+cR86aVpCXBvcjZ0Q71PJW2tfW66eRA+gCcrtOq3WW1/Z4\nLrld30Iem48HxkrajgwQnT+SPLHygvalt9WVz/XkYDSO3C/OB/6p1n0NeWw/pdbfl9yf9qr8J5L7\n96tr+YR6jCcvTV1T5T2avGy1K3nc3Exu0/HV5+PI2d2kypOO/hS5fWdWfVqz8Nbnn++s9SZJejy5\nPQRsKWlCrT+Z9qXmieRlV8j9D3L22rIluf/9H3msj6G9rZ9Q/TGJ/PhjP3Kc2oU8KfkDOdOD3H6S\ntEvV9bW0T35a++RgtiKvQq0mZ9Rjh1i3v98Df9can2vchTzW96r39iJPcJG0E7A0Ik4nL+Pu1T/D\nh1jPGxouJP+VROd77yAvC+1HXgL6AR03NdQ6rWvZ363XryV3+vvIs7IZlW417ZshDiZ37Nnkxv81\nOd1eVO9dTh6krZsajiXPJA8GTq1yLiJnGleRB8ZTyB1qeT2Wkpcy+sjLA4vJgWk+7Q8vz6/3lpAD\nwzzaPxB7Y637FzKQdX5o37pRYBY5SF1f67Zmf4vIgXUR7ctcq8mz7NaH5K0Pq2+u95dVfy2pZa2z\nyn3JmVDrctqHq213VdtXkLOJG6r8ZbRvpvh89dVZtV7nh/wzyVlg6yxxebX7o9VfrQ+RV5HX3V9M\nnmy0LvlF1edK2rOk5bRvDrm+ox6LyZON2zv6cDV56fPX1Y7bq15nkgPrA5V+Ebl/3Uv7Q/PzaF86\nubPSd974sbDKXVBpllf7V3X0wfJKdwQ5g2ndrHEtOcDMpn2DyNLK89fkwdq6XNPK52RyX5hb617H\nQ29quIz27HkVdTdq5fUC2jcCHFd5z6r3WvtOX23Xs8l/47Ky2tf6UH0uuf1bNx5cSvu4XFXbo7Ws\ndWNM60P/+bQvF7dmWZ+qbba48mgFpBXVttnkJcdZ5OdIrb5eSnt/ah0Lizva0LrisbraeWelOYkM\nuK2bMVbTPg7mVZ0eoL3fzCfHiFYfnUHuz18hj4vWvtgHvL2OgQ/RDvIrybFkWq13B7lPXUb7Zo6V\n5L4zqx63kCfPX6o8jiLHpwXkMbys0n2M9j4yhQzYd1TZ91Z/bkOezM6r7fbhWta6qeFXtf7V5Nhw\nHYPf1LAbuY/9gfzMsXUn3n7UnXr1+kTgqAFuajioyrgSOK/e24yclc8kZ+Gzq69eXGXNrL4a8g7I\n3t2+169xvU7XRb4TyUCxVa/SkYFzJrBbt+2h+1uf1yrvqnfrDplDyQ9bh8r/YPKyxPr08WB1nLG+\nea+pLPIE4MO0b319WF3q/Un191G1XXdYl3ata169aHu39WEdbz9fy7p1Vcaa1uu/nH53cW2Mj7U9\npkfjYyT+H9Kwqy+/nQJ8Ltbwodq6pqsvjZ0DnBX5QeMGs455Px04sT7Pu5+87DhY/l8iz3oG/SLk\netTxHeTlqNeta95rUVY3y8+RtDV5We5jEXHXepS1VnltCE2rjw2rro/p0co/rmpmZo3g37IzM7NG\ncEAyM7NGcEAyM7NGcEAyM7NGcEAyM7NG+H8ZrCli71KC4AAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x2a7569f2710>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#8\n",
    "\n",
    "species = penguins_df['species']\n",
    "y_pos = np.arange(len(species))\n",
    "mass = penguins_df['mass']\n",
    " \n",
    "plt.bar(y_pos, mass, align='center', alpha=0.5)\n",
    "plt.xticks(y_pos, species)\n",
    "plt.ylabel('Mass')\n",
    "plt.title('Mass for each species of penguin')\n",
    " \n",
    "plt.show()        #not getting the names evenly spread, but quite out of time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
